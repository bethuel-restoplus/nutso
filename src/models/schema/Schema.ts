import { ArraySchema } from "./ArraySchema";
import { BooleanSchema } from "./BooleanSchema";
import { DateSchema } from "./DateSchema";
import { NumberSchema } from "./NumberSchema";
import { ObjectSchema } from "./ObjectSchema";
import { StringSchema } from "./StringSchema";
import { TupleSchema } from "./TupleSchema";

// prettier-ignore
export type Schema<T, R = T> = T extends string ? StringSchema<R> :
                        T extends number ? NumberSchema :
                        T extends Date ? DateSchema :
                        T extends boolean ? BooleanSchema :
                        T extends Array<infer E> ? ArraySchema<E, R> : 
                        T extends [infer U, ...unknown[]] ? TupleSchema<U> :
                        T extends Object ? ObjectSchema<T, R> :
                        never;
