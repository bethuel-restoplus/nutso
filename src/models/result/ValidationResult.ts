import { FieldPath } from "../FieldPath";

/**
 * TODO remodel this to include the path that triggered the error message
 * For any given object, the error could have occured deep in the tree
 * So along with the error message, we should have a path that points to what triggered that particular error
 *
 * Expolore the possibility of including all the errors at any given level.
 * Think, and redesign on white board and then do it again
 */
export type ValidationResult = {
  isValid: boolean;
  errorMessage: string;
  errorPath: FieldPath;
};
