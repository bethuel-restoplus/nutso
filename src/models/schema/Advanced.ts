type False = "0";
type True = "1";
type If<C extends True | False, Then, Else> = { "0": Else; "1": Then }[C];

type Diff<T extends string | number | symbol, U extends string> = ({ [P in T]: P } & { [P in U]: never } & { [x: string]: never })[T];

type X<T> = Diff<keyof T, keyof Object>;

type Is<T, U> = (Record<X<T & U>, False> & Record<any, True>)[Diff<X<T>, X<U>>];

type DeepPartial<T> = {
  [P in keyof T]?: If<Is<Function & T[P], Function>, T[P], DeepPartial<T[P]>>;
};

type Customer = {
  name: string;
  address: {
    line1: string;
    city: string;
    props: { [key: string]: string };
  };
};

type Props = { [key: string]: string };
const props: Props = { 1: "a" };

type C1 = DeepPartial<Customer>;

const c1: C1 = {
  name: "John",
  address: {
    line1: "66 Collins St",
    props: {
      1: "foo"
    }
  }
};

type Foo = {
  type: "foo";
  foo: string;
};

type Bar = {
  type: "bar";
  bar: number;
};

type Baz = Foo | Bar;
const baz: Baz = {
  type: "foo",
  foo: "string"
};
