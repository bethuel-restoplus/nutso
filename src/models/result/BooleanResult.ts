import { ValidationResult } from "./ValidationResult";

export type BooleanResult = ValidationResult & {};
