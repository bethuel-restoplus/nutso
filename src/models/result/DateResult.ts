import { ValidationResult } from "./ValidationResult";

export type DateResult = ValidationResult & {};
