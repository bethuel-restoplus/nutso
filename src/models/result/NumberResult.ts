import { ValidationResult } from "./ValidationResult";

export type NumberResult = ValidationResult & {};
