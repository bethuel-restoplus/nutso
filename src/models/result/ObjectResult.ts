import { Result } from "./Result";
import { ValidationResult } from "./ValidationResult";

export type ObjectResult<T extends { [key: string]: any }> = ValidationResult & {
  properties: {
    [P in keyof T]: Result<T[P]>;
  };
};
