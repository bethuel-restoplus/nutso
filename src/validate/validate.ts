import { Schema } from "../models/schema/Schema";
import { Result } from "../models/result/Result";
import { validateString } from "./validateString";
import { StringSchema } from "../models/schema/StringSchema";
import { NumberSchema } from "../models/schema/NumberSchema";
import { validateNumber } from "./validateNumber";
import { BooleanSchema } from "../models/schema/BooleanSchema";
import { validateBoolean } from "./validateBoolean";
import { ObjectSchema } from "../models/schema/ObjectSchema";
import { validateObject } from "./validateObject";
import { FieldPath } from "../models/FieldPath";
import { validateArray } from "./validateArray";
import { ArraySchema } from "../models/schema/ArraySchema";
import { validateDate } from "./validateDate";
import { DateSchema } from "../models/schema/DateSchema";

/**
 * This is an internal API that takes the root as an argument
 * and called recursively from `validateObject` and `validateArray` methods
 *
 * @param o
 * @param root
 * @param schema
 */
export const _validate = <T, R>(o: T | null, root: R, schema: Schema<T, R>): Result<T> => {
  //
  if (!schema) {
    throw new Error("Schema should not be null");
  }

  /**
   * validate object against the schema
   */
  switch (schema.type) {
    case "number":
      return validateNumber(o, root, schema as NumberSchema<R>) as Result<T>;
    case "string":
      return validateString(o, root, schema as StringSchema<R>) as Result<T>;
    case "boolean":
      return validateBoolean(o, root, schema as BooleanSchema<R>) as Result<T>;
    case "array":
      return validateArray(o as any, root, schema as ArraySchema<T, R>) as Result<T>;
    case "object":
      return validateObject(o, root, schema as ObjectSchema<T, R>) as Result<T>;
    case "date":
      return validateDate(o, root, schema as DateSchema<R>) as Result<T>;
  }
  throw new Error(`Unhandled data type`);
};

export const validate = <T>(o: T, schema: Schema<T>): Result<T> => {
  return _validate(o, o, schema);
};
