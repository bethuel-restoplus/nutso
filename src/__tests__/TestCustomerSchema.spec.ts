import { Schema } from "../schema/Schema";
import { validate } from "../validate/validate";
import { Result } from "../result/Result";

type Customer = {
  name: string;
  dob: Date;
  height: number;
};

const customerSchema: Schema<Customer> = {
  type: "object",
  properties: {
    name: {
      type: "string",
      minLength: 3,
      maxLength: 24
    },
    dob: {
      type: "date"
    },
    height: {
      type: "number",
      min: 0.1,
      max: 7.5
    }
  }
};

const customer: Customer = {
  name: "John Appleseed",
  dob: new Date(), // just born
  height: 3.2
};

test(`Customer schema`, () => {
  const result: Result<Customer> = validate(customer, customerSchema);
  expect(result).toMatchSnapshot();
});
