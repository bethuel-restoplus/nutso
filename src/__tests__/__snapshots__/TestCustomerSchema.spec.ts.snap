// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Customer schema 1`] = `
Object {
  "errorMessage": "",
  "errorPath": Array [],
  "isValid": true,
  "properties": Object {
    "dob": Object {
      "errorMessage": "",
      "errorPath": Array [],
      "isValid": true,
    },
    "height": Object {
      "errorMessage": "",
      "errorPath": Array [],
      "isValid": true,
    },
    "name": Object {
      "errorMessage": "",
      "errorPath": Array [],
      "isValid": true,
    },
  },
}
`;
